File: solutions/cpp/allergies/4/allergies.cpp
#include "allergies.h"

namespace allergies {

    const std::unordered_map<std::string, int> allergy_test::allergy_map = {
            {"eggs", 1},
            {"peanuts", 2},
            {"shellfish", 4},
            {"strawberries", 8},
            {"tomatoes", 16},
            {"chocolate", 32},
            {"pollen", 64},
            {"cats", 128}
        };

    allergy_test::allergy_test(int initial_score) : score(initial_score) {}

    bool allergy_test::is_allergic_to(const std::string& substance) const {
        auto it = allergy_map.find(substance);
        return it != allergy_map.end() && score & it->second; 
    }

    std::unordered_set<std::string> allergy_test::get_allergies() {
        std::unordered_set<std::string> result;
        
        for (const auto& pair : allergy_map) {
            result.emplace(pair.first);
        }

        return result;
    }
}  // namespace allergies

File: solutions/clojure/allergies/1/src/allergies.clj
(ns allergies)

(def allergy-names [:eggs 
                    :peanuts 
                    :shellfish 
                    :strawberries
                    :tomatoes
                    :chocolate
                    :pollen
                    :cats])

(def allergy-scores (take 
                     (count allergy-names) 
                     (iterate #(bit-shift-left % 1) 1)))

(def allergy-by-score (zipmap allergy-scores allergy-names))

(defn allergies [score]
  (filter 
   (complement nil?)
   (map 
    #(allergy-by-score (bit-and % score)) 
    allergy-scores)))

(defn allergic-to? [score allergen]
  (some #(= allergen %) (allergies score)))
